### PR TITLE
3.0.2

### PR DESCRIPTION
**Pull request by tamtamchik**
This pull request implements the Lido v3.0.2 upgrade, focusing on stricter withdrawal controls for vaults with obligations shortfall and improved type safety in on-chain value assignments. It also updates deployment scripts, parameters, and configuration files to support the new release, and removes legacy upgrade logic. The most important changes are grouped below:

### Protocol Logic and Security Improvements

* In `VaultHub`, partial validator withdrawals are now strictly forbidden for vaults with obligations shortfall, regardless of the requested withdrawal amount. This closes a loophole where partial withdrawals could be used to delay required rebalancing exits. The logic is updated to check for any obligations shortfall and block partial withdrawals accordingly.
* In `LazyOracle`, the redundant check that `_maxLiabilityShares` must not exceed the on-chain `record.maxLiabilityShares` has been removed, simplifying the validation logic.
* Type safety improvements: assignments to `cumulativeLidoFees`, `minimalReserve`, and `maxLiabilityShares` in `VaultHub` now use `SafeCast` to prevent overflows. [[1]](diffhunk://#diff-b54923296b324173216e4664a344200253688e70785fe69362ffe6adbaf1edb8L1109-R1111) [[2]](diffhunk://#diff-b54923296b324173216e4664a344200253688e70785fe69362ffe6adbaf1edb8L1123-R1127)

### Deployment and Upgrade Infrastructure

* Added new deployment and mock-upgrade steps and scripts for v3.0.2, including `steps-deploy-v3-02.json`, `steps-mock-upgrade-v3-02.json`, and their associated TypeScript logic. These scripts deploy the new implementations and enforce confirmation before proceeding. [[1]](diffhunk://#diff-3eadb554d9b274d3ee781ad0b9b1f1c483c5aead742f242c419bdc977fa7ae49R1-R3) [[2]](diffhunk://#diff-64bf26928751731c8afb6f7e990853d6441b028540cbf2e3849121d4493dc2acL2-R2) [[3]](diffhunk://#diff-bb2053f6e4d99e96e18c9eeb4a8c3b8df7cdc37a17ede16fde6272696902f53cR1-R90) [[4]](diffhunk://#diff-faa4170d357342d60b7de2d3ca964ff55e5d8aaae58abce7bdd540a90ab2112aR1-R28)
* Updated deployment parameters and configuration files to use `maxRelativeShareLimitBP` for the vault hub, removing the legacy predeposit guarantee parameters. [[1]](diffhunk://#diff-2aa804122811c51054b5d57b16f76646b3fdee0e2b2a89b9c7900ceafc1c1c2bL4-R6) [[2]](diffhunk://#diff-cb7ef3fca9fb9326d03f372a404758c37779299d71d82375c5b23a8e3690e89dL4-R6)
* Updated `package.json` with new version `3.0.2`, new deployment scripts, and integration test step files for the upgrade. [[1]](diffhunk://#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519L3-R3) [[2]](diffhunk://#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519L34-R35) [[3]](diffhunk://#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519R47)

### Test and Integration Updates

* Updated and added tests to reflect the new withdrawal restrictions and validation logic, ensuring that partial withdrawals are correctly forbidden when vaults are unhealthy, and that the removed check in `LazyOracle` is no longer tested. [[1]](diffhunk://#diff-7450085f72f6495d6d2546d992e0dda67413e2b870d2e0bd082a6fc37d6eed99L877-R877) [[2]](diffhunk://#diff-7450085f72f6495d6d2546d992e0dda67413e2b870d2e0bd082a6fc37d6eed99L897-R897) [[3]](diffhunk://#diff-f08f86e7289c2abc8c6453563369605b4626c441569bc88b7842dda7fc1cea73L113-L118)
* Updated mock contracts to support new logic (`isVaultInJail`).

### Deployment Addresses

* Updated implementation addresses for `LazyOracle` and `VaultHub` in both `deployed-mainnet.json` and `deployed-hoodi.json` to point to the new v3.0.2 deployments. [[1]](diffhunk://#diff-373ae28789297e1987b7b17a33c6c2900e8790c1874f2d20aa232f6131e4700bL412-R412) [[2]](diffhunk://#diff-373ae28789297e1987b7b17a33c6c2900e8790c1874f2d20aa232f6131e4700bL739-R739) [[3]](diffhunk://#diff-226936aaf695189792c23fd3974a38c6113a119a624adc39c63733e68d522151L524-R524) [[4]](diffhunk://#diff-226936aaf695189792c23fd3974a38c6113a119a624adc39c63733e68d522151L873-R873)

### Cleanup

* Removed legacy upgrade scripts and parameters related to the predeposit guarantee, which are no longer relevant in v3.0.2.

These changes collectively harden the protocol against edge-case withdrawal exploits, improve safety, and prepare the deployment infrastructure for the v3.0.2 release.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -31,8 +31,8 @@
     "test:integration": "MODE=forking hardhat test test/integration/**/*.ts",
     "test:integration:trace": "MODE=forking hardhat test test/integration/**/*.ts --trace --disabletracer",
     "test:integration:fulltrace": "MODE=forking hardhat test test/integration/**/*.ts --fulltrace --disabletracer",
-    "test:integration:upgrade": "GAS_LIMIT=16000000 STEPS_FILE=upgrade/steps-mock-voting.json yarn test:integration:upgrade:helper test/integration/**/*.ts",
-    "test:integration:upgrade:helper": "NETWORK_STATE_FILE=deployed-mainnet.json MODE=forking UPGRADE=true GAS_PRIORITY_FEE=1 GAS_MAX_FEE=100 DEPLOYER=0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266 hardhat test --trace --disabletracer",
+    "test:integration:upgrade": "GAS_LIMIT=16000000 STEPS_FILE=upgrade/steps-mock-upgrade-v3-02.json yarn test:integration:upgrade:helper test/integration/**/*.ts",
+    "test:integration:upgrade:helper": "MODE=forking UPGRADE=true GAS_PRIORITY_FEE=1 GAS_MAX_FEE=100 DEPLOYER=0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266 hardhat test --trace --disabletracer",
```
[the remark this answers is not in the document]
Comment (dry914) [reply package.json:35]: `lib/state-file.ts` already derives the filename from the hardhat network name (e.g. `deployed-mainnet.json, deployed-hoodi.json`), so hardcoding it here was redundant and broke Hoodi fork tests in CI

**Review by dry914 — APPROVED** [review]
LGTM

**Review by Ziars — APPROVED** [review]
